Moves variorum_log to an environment variable.

chaz-browne opened the pull request:
# Description

Moves VARIORUM_LOG from a CMake build variable to an environment variable. This enables logging to be enabled/disabled at runtime.

To enable logging:
```
$ VARIORUM_LOG=1 ./examples/variorum-print-power-example 
_LOG_VARIORUM_ENTER:/g/g19/brink2/git_root_pub/variorum-dev/src/variorum/variorum.c:variorum_print_power::598
Running volta_get_power
_NVIDIA_GPU_POWER_USAGE Host Socket DeviceID Power_W
_NVIDIA_GPU_POWER_USAGE lassen32 0 0 34.312002
_NVIDIA_GPU_POWER_USAGE lassen32 0 1 37.793002
_NVIDIA_GPU_POWER_USAGE lassen32 1 2 34.809002
_NVIDIA_GPU_POWER_USAGE lassen32 1 3 34.832002
_LOG_VARIORUM_EXIT:/g/g19/brink2/git_root_pub/variorum-dev/src/variorum/variorum.c:variorum_print_power::618
```

To disable logging:
```
$ VARIORUM_LOG=0 ./examples/variorum-print-power-example 
Running volta_get_power
_NVIDIA_GPU_POWER_USAGE Host Socket DeviceID Power_W
_NVIDIA_GPU_POWER_USAGE lassen32 0 0 34.764002
_NVIDIA_GPU_POWER_USAGE lassen32 0 1 37.720002
_NVIDIA_GPU_POWER_USAGE lassen32 1 2 34.764002
_NVIDIA_GPU_POWER_USAGE lassen32 1 3 34.764002
```

`$ ./examples/variorum-print-power-example` is equivalent to the above.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Build/CI update

# How Has This Been Tested?

- [x]  Comus (Broadwell),
- [x]  Thompson (Haswell)
- [x]   Alehouse (Power9)
- [x]   Rhetoric (Skylake, single socket)
- [x]   ARM Juno r2
- [x]   Corona AMD GPU (fixed by using ROCM 5.2.0)
- [x]   Lassen with the CUDA host-config file build for NVML testing (different than alehouse)

# Checklist:

- [x] I have run `./scripts/check-code-format.sh` and confirm my C/C++ code follows the style guidelines of variorum
- [x] I have run `flake8` and `black --check --diff` and confirm my python code follows the style guidelines of variorum
- [x] I have run `./scripts/check-rst-format.sh` and confirm my documentation follows the style guidelines of variorum
- [x] I have added comments in my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass with my changes

chaz-browne committed:
WIP, changing compile time log to runtime.

Set the VARIORUM_LOG environment variable in order to get log
messages.
This commit fixes variorum_enter();.

chaz-browne committed:
WIP, continue converting logs.

chaz-browne committed:
Replaces #VARIORUM_LOG with environment variable.

chaz-browne committed:
Style fixes.

tpatki commented:
@chaz-browne @rountree : 

Build fails with a missing `string.h` in some files. I could find at least one place (IvyBridge files) where it needs to be added, there may be other files where you need to include `string.h`. 

Also, in many places I think you have a copy-paste error. For instance, on Skylake nodes, the build fails with:

```
/home/runner/work/variorum/variorum/src/variorum/Intel/Skylake_55.c:427:35: error: ordered comparison of pointer with integer zero [-Werror=extra]
[364](https://github.com/LLNL/variorum/runs/7079250509?check_suite_focus=true#step:5:365)
     if(val != NULL && getenv(val) > 0){ 
```

I think you meant to check for `if (val != NULL && strlen(val) > 0)` -- not an `getenv(val) > 0`. `Getenv` returns a `char*` so you can't make an integer based comparison. 

-----

You also put the closing brace for the `fm_06_4f_get_node_power_json(json_t *get_power_obj)` in Broadwell_4F.c in the wrong place, leaving a stray call to `json_get_power_data` outside the function. 

-----

You have a  similar error here, where the call to `son_get_power_domain_info(get_domain_obj, msrs.msr_pkg_power_info,` was deleted incorrectly when adding `VARIOUM_LOG`.
https://github.com/LLNL/variorum/pull/270/files#diff-9eb8cb80e2412cc9674989480c818658b0cefc7e5545c5229ecd4a859161494aL440

https://github.com/LLNL/variorum/pull/270/files#diff-9eb8cb80e2412cc9674989480c818658b0cefc7e5545c5229ecd4a859161494aR442

There may be a few others that I haven't caught.

----
Also, as a general note:  before submitting any PRs, please run `run scripts/check-code-format.sh` so it formats your code correctly and doesn't fail.

tpatki commented on src/variorum/Intel/IvyBridge_3E.c:
```diff
@@ -77,10 +77,11 @@ int fm_06_3e_get_power_limits(int long_ver)
-#endif
-
+    char *val = getenv("VARIORUM_LOG");
+    if (val != NULL && strlen(val) > 0)
```
Add string.h header for this file.

tpatki commented on src/variorum/Intel/Skylake_55.c:
```diff
@@ -72,10 +73,11 @@ int fm_06_55_get_power_limits(int long_ver)
-#endif
-
+    char *val = getenv("VARIORUM_LOG");
+    if (val != NULL && getenv(val) > 0)
```
Should be `strlen(val) > 0`

tpatki commented on src/variorum/Intel/Skylake_55.c:
```diff
@@ -132,10 +134,11 @@ int fm_06_55_cap_power_limits(int package_power_limit)
-#endif
-
+    char *val = getenv("VARIORUM_LOG");
+    if (val != NULL && getenv(val) > 0)
```
Should be `strlen(val) > 0`

tpatki replied:
This happens in many places in this file, please fix.

tpatki commented on src/variorum/Intel/Skylake_55.c:
```diff
@@ -246,10 +251,11 @@ int fm_06_55_get_features(void)
-#endif
-
+    char *val = getenv("VARIORUM_LOG");
+    if (val != NULL && getenv(val) > 0)
```
Should be strlen(val) > 0

tpatki commented on src/variorum/Intel/Skylake_55.c:
```diff
@@ -146,9 +149,11 @@ int fm_06_55_cap_power_limits(int package_power_limit)
-    printf("Running %s\n", __FUNCTION__);
-#endif
+    char *val = getenv("VARIORUM_LOG");
+    if (val != NULL && getenv(val) > 0)
```
Should be strlen(val) > 0

tpatki commented on src/variorum/Intel/Broadwell_4F.c:
```diff
@@ -439,23 +462,26 @@ int fm_06_4f_get_node_power_json(json_t *get_power_obj)
+        printf("Running %s\n", __FUNCTION__);
+    }
+    msrs.msr_dram_power_info,
+    msrs.msr_rapl_power_unit, msrs.msr_pkg_power_limit;
```
Incorrectly deleted the original call to ` json_get_power_domain_info`, stray arguments left behind.

rountree commented:
@tpatki Testing across those architectures goes beyond what I was hoping to do here.  I had intended this PR to get command line parsing into `getopt()` and fix the existing help messages (e.g., `--help` was mentioned but often not supported).

I think it would be valuable to add architecture-specific messages to the `-h` output, but I think that's a separate PR.

If you'd like to see that feature go in this summer I'm sure @chaz-browne can handle it, at which point this becomes a question of bookkeeping.

Thoughts?

rountree-alt committed:
VARIORUM_LOG rebase.

Removed VARIORUM_LOG from cmake files, documentation, and
configuration files.  Checked formatting.  Need to verify if
building documentation works.

rountree-alt committed:
Rebase to dev, remove conflicts arising from AMD GPU port, and format files.

VARIORUM_LOG cleanup pass.

Removed VARIORUM_LOG from cmake files, documentation, and
configuration files.  Checked formatting.  Need to verify if
building documentation works.

tpatki commented:
@slabasan Why aren't CI checks triggered automatically on this PR? I couldn't find the option to start CI tests either...I must be missing something simple here.

slabasan commented:
Going to re-open a new PR in hopes CI launches

rountree committed:
Adds missing #include <string.h>

rountree committed:
Fixes coding typo.

tpatki committed:
Rebase and resolve conflicts on this PR.

Nvidia port cleanup (#298)

* Nvidia port internal cleanup, change VARIORUM_WITH_NVIDIA to VARIORUM_WITH_NVIDIA_GPU

* separate local headers from standard headers

Co-authored-by: Stephanie Brink <brink2@llnl.gov>

rountree commented:
I'm working on it....

rountree committed:
Fixes getenv/strlen hiccup.

tpatki commented:
@slabasan @rountree I rebased and resolved some conflicts here, but this is not ready to merge as we need to update the VARIORUM_LOG part in the new AMD GPU port that was merged today. I will get to it tonight or tomorrow.

tpatki commented:
Ugh, I think I messed something up in the rebase. Sorry. Fixing now.

@slabasan @rountree

tpatki committed:
Fixing rebase conflicts...attempt 1

tpatki committed:
Fix Rebase conflicts

tpatki committed:
Remove VARIORUM_LOG from remaining host-config files

tpatki committed:
Fix two places in Intel/Intel_06_6A.c where the VARIORUM_LOG change wasn't made

tpatki committed:
Add VARIORUM_LOG env var to AMD

tpatki committed:
Format files.

tpatki committed:
Add missing string.h in places that need strlen

tpatki committed:
Fix build errors on NVIDIA strnlen --> strlen

tpatki commented:
@rountree @slabasan @chaz-browne 

I found a problem when testing, Stephanie and I will fix it tomorrow -- don't worry about it. 

Just documenting it here, so we remember to fix this and test it on all known architectures. 

Basically, you are not checking for the value of VARIORUM_LOG, you are only checking for its length to be greater than zero. That means, if someone sets that variable to `0`, it will still print the log enter and exit msgs, which should not be the case. For example, see the result below from the CUDA build on Lassen:

```
[patki1@lassen88:examples]$ export VARIORUM_LOG=0

[patki1@lassen88:examples]$ ./variorum-print-power-example 
_LOG_VARIORUM_ENTER:/g/g90/patki1/src/var-clean/chaz/variorum/src/variorum/variorum.c:variorum_print_power::598
Running volta_get_power
_NVIDIA_GPU_POWER_USAGE Host Socket DeviceID Power_W
_NVIDIA_GPU_POWER_USAGE lassen88 0 0 34.807002
_NVIDIA_GPU_POWER_USAGE lassen88 0 1 35.787002
_NVIDIA_GPU_POWER_USAGE lassen88 1 2 36.768002
_NVIDIA_GPU_POWER_USAGE lassen88 1 3 35.810002
_LOG_VARIORUM_EXIT:/g/g90/patki1/src/var-clean/chaz/variorum/src/variorum/variorum.c:variorum_print_power::618
_LOG_VARIORUM_ENTER:/g/g90/patki1/src/var-clean/chaz/variorum/src/variorum/variorum.c:variorum_print_power::598
Running volta_get_power
_NVIDIA_GPU_POWER_USAGE lassen88 0 0 35.297002
_NVIDIA_GPU_POWER_USAGE lassen88 0 1 35.787002
_NVIDIA_GPU_POWER_USAGE lassen88 1 2 36.768002
_NVIDIA_GPU_POWER_USAGE lassen88 1 3 35.810002
_LOG_VARIORUM_EXIT:/g/g90/patki1/src/var-clean/chaz/variorum/src/variorum/variorum.c:variorum_print_power::618
```

It should be checking for `if (val != NULL && atoi(*val) > 0)`  or something like that instead of `if (val != NULL && strlen(val) > 0)`, which is what you currently have. 


ToDo: Test the PR build on:

- [ ]  Comus (Broadwell),
- [ ]   Thompson (Haswell),
- [ ]   Alehouse (Power9)
- [ ]   Rhetoric (Skylake, single socket)
- [ ]   ARM Juno r2
- [ ]   Corona AMG GPU
- [ ]   Lassen with the CUDA host-config file build for NVML testing (different than alehouse)

.

slabasan committed:
check if VARIORUM_LOG=1 when enabling log statements

slabasan committed:
update docs

slabasan committed:
remove unused string.h header

strlen replaced with atoi

slabasan committed:
cleanup, restore newlines

slabasan committed:
remove added newlines in variorum.c

slabasan committed:
restoring headers

tpatki committed:
Add string.h back to the ARM port as it needs it

tpatki reviewed (dismissed):
LGTM

slabasan committed:
small fixes

slabasan committed:
small fix

slabasan committed:
move string.h to original location in ARM

slabasan committed:
add new lines
